<div align="center">
  <h2 style="color:#1a6eb5; border-bottom: 2px solid #1a6eb5; padding-bottom: 6px;">
    Gestión Cuantitativa de Riesgo de Crédito
  </h2>
  <h4 style="color:#555; margin-top:4px;">Construcción Modelo Scoring de Crédito</h4>
  <p style="font-size:14px;">
    <strong>Universidad de Chile</strong>  | 
    Magíster en Analítica Financiera  | 
    <em>Analítica Financiera</em>
  </p>
  <p style="color:#1a6eb5; font-weight:bold; font-size:13px;">Franco Mansilla Ibáñez</p>
  <br>
  <a href="https://www.francomansilla.com">
    <img src="https://img.shields.io/badge/Web-francomansilla.com-1a6eb5?style=flat-square&logo=globe"/>
  </a>
   
  <img src="https://img.shields.io/badge/pandas-2.x-150458?style=flat-square&logo=pandas&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/numpy-1.x-013243?style=flat-square&logo=numpy&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/scipy-1.x-8CAAE6?style=flat-square&logo=scipy&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/tabulate-0.9-4CAF50?style=flat-square"/>
   
  <img src="https://img.shields.io/badge/plotly-5.x-3F4F75?style=flat-square&logo=plotly&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/statsmodels-0.14-1a6eb5?style=flat-square"/>
   
  <img src="https://img.shields.io/badge/scikit--learn-1.x-F7931E?style=flat-square&logo=scikit-learn&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/mrmr__classif-0.2-8A2BE2?style=flat-square"/>
   
  <a href="https://arxiv.org/abs/1908.05376">
    <img src="https://img.shields.io/badge/Paper-mRMR-red?style=flat-square&logo=arxiv&logoColor=white"/>
  </a>
</div>

  <h2 style="color:#1a6eb5; border-bottom: 2px solid #1a6eb5; padding-bottom: 6px;">


# 1. Instancias 

* Control de verisón de librerías.

In [2]:
#%pip install --upgrade --force-reinstall --no-cache-dir pandas==2.2.2 numpy==1.26.4 scipy==1.13.1 tabulate==0.9.0 plotly==5.23.0 statsmodels==0.14.2 mrmr-selection==0.2.8 scikit-learn==1.5.1
!pip install pandas==2.2.2 numpy==1.26.4 scipy==1.13.1 tabulate==0.9.0 plotly==5.23.0 statsmodels==0.14.2 mrmr-selection==0.2.8 scikit-learn==1.5.1

Looking in indexes: https://fmansii%40bci.cl:****@somosmach.jfrog.io/artifactory/api/pypi/data-analytics-pypi/simple
    python-dotenv (>=0.13.*) ; extra == 'standard'
                   ~~~~~~~^

[notice] A new release of pip is available: 25.2 -> 26.0.1
[notice] To update, run: pip install --upgrade pip


* Importar librerías necesarias.

In [102]:
# Librerías Base de Datos
import pandas as pd
import numpy as np

# Estadística
from scipy.stats import levene
from scipy import stats
from tabulate import tabulate

# Visualización
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import plotly.express as px
import plotly.figure_factory as ff

# Preprocesamiento - Modelamiento
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from mrmr import mrmr_classif
from sklearn.metrics import roc_curve, f1_score, roc_auc_score
from sklearn.base import is_classifier
from sklearn.ensemble import RandomForestClassifier

---
# 2. Base de Datos

En esta sección revisaremos:

* 1. Carga de Base de Datos.
* 2. Split Sample. separamos la muestra en entrenamiento (70%) y validación (30%).

---

**1. Carga de Base de Datos**

In [103]:
file_path = '/Users/francomansilla/Library/CloudStorage/GoogleDrive-franco.andres.mansilla@gmail.com/Mi unidad/universidad de chile/Magister Analitica Negocio/Analítica Financiera v2/3. Riesgo Crédito/Código Python v2/'

df = pd.read_excel(file_path + 'Base de Datos Python.xlsx', sheet_name='4. BD a Modelar')

print("Dimensión", df.shape)
df

Dimensión (500, 13)


,id,fecha_corte,default,edad,n_educ,a_emp,a_dir,ing_h,deu_ing,deu_tc,deu_o,flag_casado,a_exp
0,880,2023-02-01,0,43,1.0,23,19,72.0,7.6,1.18,4.29,1.0,NaN
1,979,2024-02-01,0,39,1.0,6,9,61.0,5.7,0.56,2.91,1.0,36.0
2,898,2023-12-01,0,34,NaN,7,15,40.0,6.4,0.95,1.61,1.0,26.0
3,767,2023-08-01,1,40,3.0,5,3,0.0,16.0,NaN,0.00,0.0,41.0
4,915,2023-08-01,1,21,2.0,2,0,20.0,4.5,0.29,0.61,1.0,3.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
495,821,2023-08-01,0,32,2.0,7,10,32.0,4.3,0.43,0.94,0.0,3.0
496,865,2023-03-01,0,50,1.0,8,27,47.0,6.9,0.40,2.84,0.0,29.0
497,659,2023-07-01,0,43,2.0,16,10,83.0,4.1,0.26,3.14,1.0,21.0
498,104,2024-05-01,1,31,1.0,1,6,21.0,2.5,0.28,0.25,NaN,NaN


**2. Split Sample**

In [104]:
df_2023 = df[df['fecha_corte'].dt.year == 2023].drop(columns=['id','fecha_corte'])
df_test  = df[df['fecha_corte'].dt.year == 2024].drop(columns=['id','fecha_corte'])

df_ent, df_val = train_test_split(df_2023, test_size=0.3, random_state=42)

print(f"Entrenamiento: {len(df_ent)} filas ({len(df_ent)/len(df_2023)*100:.1f}% de 2023)")
print(f"Validación:    {len(df_val)} filas ({len(df_val)/len(df_2023)*100:.1f}% de 2023)")
print(f"Test:          {len(df_test)} filas (100% de 2024)")

Entrenamiento: 233 filas (70.0% de 2023)
Validación:    100 filas (30.0% de 2023)
Test:          167 filas (100% de 2024)


---- 
# 3. Análisis Calidad de la Información (ACI)

* 1. Identificación y Tratamiento de Valores Atípicos 
* 2. Identificación y Tratamiento de Missing Values
* 3. Índice de Estabilidad de las Variables (CSI) e Information Value (IV)

----

**1.1. ***Identifcación*** variables con datos atípicos**

La idea es usar los estadisticos para poder conocer la base de datos y detectar:
1. **Concentración:** usando los percentiles, por ejemplo, si p5% = p95% entonces la variable tiene una alta concentración + Asimetría.
2. **Datos atípicos:** y frecuencia usando las medidas combinada con mediana + Curtosis. 

In [105]:
def estadistica_descriptiva(df):
    df_num = df.select_dtypes(include='number')
    
    # Estadísticas numéricas
    stats = df_num.describe(percentiles=[0.01, 0.05, 0.5, 0.95, 0.99]).T
    stats = stats[['min', '1%', '5%', 'mean', '50%', 'std', '95%', '99%', 'max']]
    stats.columns = ['Mínimo', 'P1', 'P5', 'Promedio', 'Mediana', 'Desv. Estándar', 'P95', 'P99', 'Máximo']
    stats['Asimetría'] = df_num.skew()
    stats['Curtosis'] = df_num.kurt()
    stats['Variación %'] = ((stats['Promedio'] - stats['Mediana']) / stats['Promedio']).round(4)
    stats['Num. Únicos'] = df_num.nunique()

    return stats.T

resultado = estadistica_descriptiva(df_ent)
display(resultado)

,default,edad,n_educ,a_emp,a_dir,ing_h,deu_ing,deu_tc,deu_o,flag_casado,a_exp
Mínimo,0.000000,21.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
P1,0.000000,21.320000,1.000000,0.000000,0.000000,0.000000,0.464000,0.000000,0.000000,0.000000,0.000000
P5,0.000000,24.000000,1.000000,0.000000,0.000000,15.250000,1.580000,0.090000,0.178000,0.000000,3.000000
Promedio,0.270386,34.978541,1.685841,8.729614,8.180258,39.446903,10.290987,1.255877,2.596943,0.516854,25.545977
Mediana,0.000000,34.000000,1.000000,8.000000,7.000000,31.500000,8.900000,0.855000,1.880000,1.000000,26.000000
Desv. Estándar,0.445116,8.041083,0.976831,6.775325,6.958754,22.963146,6.584966,1.197272,2.291593,0.501126,14.103894
P95,1.000000,50.000000,3.000000,22.000000,23.000000,82.750000,23.100000,3.802000,7.820000,1.000000,47.000000
P99,1.000000,52.680000,5.750000,25.680000,26.000000,110.250000,26.436000,4.735700,9.675200,1.000000,49.000000
Máximo,1.000000,56.000000,6.000000,27.000000,31.000000,116.000000,30.600000,5.250000,9.970000,1.000000,50.000000
Asimetría,1.040636,0.430233,1.906783,0.653788,0.981379,1.066905,0.726043,1.352840,1.404316,-0.068029,-0.073775


In [106]:
vars_alta_variacion = resultado.loc['Variación %'][resultado.loc['Variación %'].abs() > 0.15].index.tolist()
print(f"Variables con variación > 15%: {vars_alta_variacion}")

Variables con variación > 15%: ['default', 'n_educ', 'ing_h', 'deu_tc', 'deu_o', 'flag_casado']


- 1.2. **Tratamiento** variables con datos atípicos

La winzorización es una técnica de tratamiento de datos atípicos que consiste en limitar los valores extremos a un rango específico. El método de establecer el piso y el techo se puede usar varios métodos: `percentiles`, `teorema de Chebyshev` entre otros. El umbral es un 15\% como máximo.

In [107]:
vars_alta_variacion = ['ing_h', 'deu_tc', 'deu_o']

In [108]:
def winsorizar(df_ent, df_val, df_test, variables, lower=0.01, upper=0.99):

    df_ent  = df_ent.copy()
    df_val  = df_val.copy()
    df_test = df_test.copy()

    for var in variables:
        p1  = df_ent[var].quantile(lower)
        p99 = df_ent[var].quantile(upper)
        df_ent[var]  = df_ent[var].clip(lower=p1, upper=p99)
        df_val[var]  = df_val[var].clip(lower=p1, upper=p99)
        df_test[var] = df_test[var].clip(lower=p1, upper=p99)
    return df_ent, df_val, df_test

# Ejecución función de winsorización
df_ent, df_val, df_test = winsorizar(df_ent, df_val, df_test, vars_alta_variacion)

print("--- Entrenamiento ---")
display(estadistica_descriptiva(df_ent[vars_alta_variacion]))

print("--- Validación ---")
display(estadistica_descriptiva(df_val[vars_alta_variacion]))

print("--- Test ---")
display(estadistica_descriptiva(df_test[vars_alta_variacion]))

--- Entrenamiento ---


,ing_h,deu_tc,deu_o
Mínimo,0.000000,0.000000,0.000000
P1,0.000000,0.000000,0.000000
P5,15.250000,0.090000,0.178000
Promedio,39.397124,1.252356,2.595177
Mediana,31.500000,0.855000,1.880000
Desv. Estándar,22.803708,1.186320,2.286018
P95,82.750000,3.802000,7.820000
P99,108.187500,4.717961,9.642944
Máximo,110.250000,4.735700,9.675200
Asimetría,1.028110,1.312277,1.393699


--- Validación ---


,ing_h,deu_tc,deu_o
Mínimo,0.000000,0.000000,0.000000
P1,0.000000,0.000000,0.000000
P5,17.000000,0.098000,0.426000
Promedio,39.134021,1.087229,2.359485
Mediana,32.000000,0.780000,1.740000
Desv. Estándar,21.221367,0.999565,1.987596
P95,82.000000,3.194000,6.582000
P99,98.080000,4.171514,8.213600
Máximo,100.000000,4.735700,9.500000
Asimetría,0.990147,1.574396,1.505305


--- Test ---


,ing_h,deu_tc,deu_o
Mínimo,0.000000,0.000000,0.000000
P1,9.100000,0.032000,0.102800
P5,17.000000,0.090000,0.350000
Promedio,39.073795,1.100532,2.430395
Mediana,33.500000,0.780000,1.690000
Desv. Estándar,21.203296,1.045115,2.136527
P95,84.750000,3.030000,7.654000
P99,102.400000,4.584000,9.264400
Máximo,110.250000,4.735700,9.675200
Asimetría,1.322861,1.539437,1.550224


- 2.1. **Identifcación** variables con missing values

**Nota.** En este caso, calcules el porcentaje de missing value con respecto a toda la base de datos (**no sobre la df_ent**), dado que por coincidencia del `split sample` pudieron quedar concentrados en el df_ent.


In [109]:
def missing_values(df):
    missing = pd.DataFrame({
        'Cant. Missing': df.isnull().sum(),
        '% Missing': (df.isnull().sum() / len(df) * 100).round(2)
    })
    missing = missing[missing['Cant. Missing'] > 0].sort_values('% Missing', ascending=False)
    return missing

missing_values(df) # Recuerda usar df original para calcular el porcentaje de missing sobre toda la base

,Cant. Missing,% Missing
a_exp,113,22.6
flag_casado,102,20.4
n_educ,13,2.6
deu_tc,12,2.4
ing_h,11,2.2
deu_o,9,1.8


- 2.2. **Tratamiento** variables con missing values

La idea es imputar la variables con missing value respetando la naturaleza de ella. 

* **Si la variables tiene +15\%** de missing value eliminala de la base de datos, pero antes dicotomomízala para conservar la información de que el valor es missing (1: con información, 0: sin información).

* **Si la variables tiene <15\%** de missing value, imputala con la promedio (si es numérica) o con la moda+1 (si es categórica).

In [110]:
def imputar_missing(df_ent, df_val, df_test, umbral=0.15):

    df_ent  = df_ent.copy()
    df_val  = df_val.copy()
    df_test = df_test.copy()
    missing_pct = df.isnull().sum() / len(df) # OJO: siempre sobre el df original.
    vars_missing = missing_pct[missing_pct > 0].index.tolist()
    vars_drop    = [v for v in vars_missing if missing_pct[v] >  umbral]
    vars_imputar = [v for v in vars_missing if missing_pct[v] <= umbral]

    # --- Variables con missing > 15%: dummy + drop ---
    for var in vars_drop:
        df_ent[f'{var}_missing']  = df_ent[var].isnull().astype(int)
        df_val[f'{var}_missing']  = df_val[var].isnull().astype(int)
        df_test[f'{var}_missing'] = df_test[var].isnull().astype(int)
    df_ent.drop(columns=vars_drop, inplace=True)
    df_val.drop(columns=vars_drop, inplace=True)
    df_test.drop(columns=vars_drop, inplace=True)
    print(f"Variables eliminadas (missing > {umbral*100:.0f}%): {vars_drop}")
    
    # --- Variables con missing <= 15%: imputación ---
    for var in vars_imputar:
        if df_ent[var].dtype == 'float64':
            valor = df_ent[var].mean()
            df_ent[var]  = df_ent[var].fillna(valor)
            df_val[var]  = df_val[var].fillna(valor)
            df_test[var] = df_test[var].fillna(valor)
            print(f"{var} (float) imputado con promedio: {valor:.4f}")
        elif df_ent[var].dtype == 'int64':
            valor = df_ent[var].max() + 1
            df_ent[var]  = df_ent[var].fillna(valor)
            df_val[var]  = df_val[var].fillna(valor)
            df_test[var] = df_test[var].fillna(valor)
            print(f"{var} (int) imputado con categoría faltante: {valor}")
    return df_ent, df_val, df_test

df_ent, df_val, df_test = imputar_missing(df_ent, df_val, df_test)

Variables eliminadas (missing > 15%): ['flag_casado', 'a_exp']
n_educ (float) imputado con promedio: 1.6858
ing_h (float) imputado con promedio: 39.3971
deu_tc (float) imputado con promedio: 1.2524
deu_o (float) imputado con promedio: 2.5952


**3. Índice de Estabilidad de las Variables (CSI) e Information Value (IV)**

El **CSI** es una medida que se utiliza para evaluar la estabilidad de una variable a lo largo del tiempo o entre diferentes muestras. Se calcula comparando la distribución de la variable en dos conjuntos de datos, como el conjunto de entrenamiento y el conjunto de validación. 

La formula para calcular el CSI es la siguiente: 
$$CSI = \sum_{i=1}^{n} \frac{(P_{i,train} - P_{i,val})^2}{P_{i,train}}$$

Los rangos de estabilidad son:
- <0.10 → Estable
- 0.10 - 0.25 → Alerta
- '>0.25 → Inestable

El **IV**, por otro lado, es una medida que se utiliza para evaluar la capacidad predictiva de una variable en relación con una variable objetivo (target). Se calcula utilizando la distribución de la variable en relación con la variable objetivo.

La formula para calcular el IV es la siguiente:
$$IV = \sum_{i=1}^{n} (P_{i,good} - P_{i,bad}) \cdot \ln\left(\frac{P_{i,good}}{P_{i,bad}}\right)$$

Los rangos de poder predictivo del IV son:
- <0.02 → Sin poder
- 0.02 - 0.10 → Débil
- 0.10 - 0.30 → Moderado
- '>.30 → Fuerte

In [111]:
def calcular_csi_iv(df_ent, df_val, df_test, variables, target, bins=10):
    resultados = []
    for var in variables:
        if var == target:
            continue
        # Crear bins con df_ent
        _, bin_edges = pd.cut(df_ent[var], bins=bins, retbins=True, duplicates='drop')
        bin_edges[0]  = -np.inf
        bin_edges[-1] =  np.inf
        # Distribución entrenamiento
        dist_ent  = pd.cut(df_ent[var],  bins=bin_edges).value_counts(normalize=True).sort_index()
        dist_val  = pd.cut(df_val[var],  bins=bin_edges).value_counts(normalize=True).sort_index()
        dist_test = pd.cut(df_test[var], bins=bin_edges).value_counts(normalize=True).sort_index()
        # CSI val
        dist_v = pd.DataFrame({'ent': dist_ent, 'val': dist_val}).fillna(0)
        dist_v['ent'] = dist_v['ent'].replace(0, 0.0001)
        dist_v['val'] = dist_v['val'].replace(0, 0.0001)
        dist_v['csi'] = (dist_v['val'] - dist_v['ent']) * np.log(dist_v['val'] / dist_v['ent'])
        csi_val = dist_v['csi'].sum()
        # CSI test
        dist_t = pd.DataFrame({'ent': dist_ent, 'test': dist_test}).fillna(0)
        dist_t['ent']  = dist_t['ent'].replace(0, 0.0001)
        dist_t['test'] = dist_t['test'].replace(0, 0.0001)
        dist_t['csi']  = (dist_t['test'] - dist_t['ent']) * np.log(dist_t['test'] / dist_t['ent'])
        csi_test = dist_t['csi'].sum()
        # Information Value (calculado sobre df_ent)
        df_bin = df_ent.copy()
        df_bin['bin'] = pd.cut(df_bin[var], bins=bin_edges)
        grouped = df_bin.groupby('bin')[target].agg(['sum', 'count'])
        grouped.columns = ['eventos', 'total']
        grouped['no_eventos'] = grouped['total'] - grouped['eventos']
        total_ev  = grouped['eventos'].sum()
        total_nev = grouped['no_eventos'].sum()
        grouped['dist_ev']  = grouped['eventos'].replace(0, 0.0001)  / total_ev
        grouped['dist_nev'] = grouped['no_eventos'].replace(0, 0.0001) / total_nev
        grouped['woe'] = np.log(grouped['dist_ev'] / grouped['dist_nev'])
        grouped['iv']  = (grouped['dist_ev'] - grouped['dist_nev']) * grouped['woe']
        iv_total = grouped['iv'].sum()
        resultados.append({
            'Variable': var,
            'CSI_val':  round(csi_val,  4),
            'CSI_test': round(csi_test, 4),
            'IV':       round(iv_total, 4)
        })
    resultado = pd.DataFrame(resultados).sort_values('IV', ascending=False)
    def estabilidad(x):
        return 'Estable' if x < 0.1 else ('Alerta' if x < 0.25 else 'Inestable')
    resultado['Estabilidad_val']  = resultado['CSI_val'].apply(estabilidad)
    resultado['Estabilidad_test'] = resultado['CSI_test'].apply(estabilidad)
    resultado['Poder Predictivo'] = resultado['IV'].apply(
        lambda x: 'Sin poder' if x < 0.02 else ('Débil' if x < 0.1 else ('Moderado' if x < 0.3 else 'Fuerte'))
    )
    return resultado

In [112]:
csi_iv_resultado = calcular_csi_iv(
    df_ent, df_val, df_test,
    df_ent.select_dtypes(include='number').columns.tolist(),
    target='default'
)
display(csi_iv_resultado)

,Variable,CSI_val,CSI_test,IV,Estabilidad_val,Estabilidad_test,Poder Predictivo
2,a_emp,0.2081,0.0847,1.8657,Alerta,Estable,Fuerte
5,deu_ing,0.1402,0.0713,0.8710,Alerta,Estable,Fuerte
3,a_dir,0.3145,0.0923,0.6856,Inestable,Estable,Fuerte
4,ing_h,0.0762,0.0502,0.5794,Estable,Estable,Fuerte
6,deu_tc,0.1394,0.1678,0.3159,Alerta,Alerta,Fuerte
0,edad,0.0646,0.1248,0.2294,Estable,Alerta,Moderado
1,n_educ,0.1415,0.0702,0.2153,Alerta,Estable,Moderado
7,deu_o,0.1057,0.0259,0.0995,Alerta,Estable,Débil
8,flag_casado_missing,0.0368,0.0153,0.0116,Estable,Estable,Sin poder
9,a_exp_missing,0.0318,0.0079,0.0027,Estable,Estable,Sin poder


**Averiguar:** ¿a qué se puede deber que algunas variables tienen la estabilidad de test sea **estable** pero en validación sea **alerta** o **inestable**? ¿Tamaño de la muestra?; ¿Estacionalidad de los periodos? ¿Efecto split de la muestra? 

----
# 4. Análisis pre-eliminar

* Siempre es importante conocer las variables en terminos de poder predictivo antes de hacer cualquier modelelo estadístico o de machine learning, para esto se puede usar técnicas univariadas como: ***Information Value*** (IV), **análisis visual** y **pruebas estadísticas**.

* **Limitación**: es un análisis univariado, no considera la interacción entre variables. Pero funciona como mecanismo de eliminación de variables.

En esta sección revisaremos:

* 1. Discriminación visual de las variables.
* 2. Discriminación estadística de las variables.

----

**1. Discriminación Visual**

* Una forma de analizar la capacidad predictiva de las variables es a través de gráficos, permiten visualizar la distribución de las variables en relación con la variable objetivo (target).

In [113]:
grupo_default    = df[df['default'] == 1]
grupo_no_default = df[df['default'] == 0]

fig = make_subplots(rows=1, cols=2, subplot_titles=['deu_tc (no discrimina)', 'deu_ing (discrimina)'])

# --- deu_tc ---
fig.add_trace(go.Histogram(x=grupo_default['deu_tc'],    nbinsx=20, opacity=0.5, name='Default',    histnorm='percent', marker_color='red',  legendgroup='default'),    row=1, col=1)
fig.add_trace(go.Histogram(x=grupo_no_default['deu_tc'], nbinsx=20, opacity=0.5, name='No Default', histnorm='percent', marker_color='blue', legendgroup='no_default'), row=1, col=1)

# --- deu_ing ---
fig.add_trace(go.Histogram(x=grupo_default['deu_ing'],    nbinsx=20, opacity=0.5, name='Default',    histnorm='percent', marker_color='red',  legendgroup='default',    showlegend=False), row=1, col=2)
fig.add_trace(go.Histogram(x=grupo_no_default['deu_ing'], nbinsx=20, opacity=0.5, name='No Default', histnorm='percent', marker_color='blue', legendgroup='no_default', showlegend=False), row=1, col=2)
fig.update_layout(
    barmode='overlay',
    title='Diferencias entre variables que discriminan la variables objetivo (y = default)',
    paper_bgcolor='rgba(255,255,255,1)',
    plot_bgcolor='rgba(255,255,255,1)',
    width=1200,
    height=600
)
fig.update_yaxes(title_text='Frecuencia Relativa (%)', col=1)
fig.update_yaxes(title_text='Frecuencia Relativa (%)', col=2)
fig.show()

**2. Discriminación Estadística**

* Se evaluar la discriminación de las variables sobre la variable objetivo (`default`), se puede usar las pruebas estadística para determinar mediante un nivel de confianza la variable es capaz de discriminar entre los buenos y malos pagadores.

* Para comparar las medias/medianas de la variable según el grupo (`default`), primero es necesario determinar si las varianzas de ambos grupos son iguales o diferentes. Para ello, se utiliza la prueba de varianza de **Fisher**.

In [114]:
def sign_estadistica(df, var_grupo, alpha_varianza=0.05):
    rs_analisis = []
    cat_0 = df[var_grupo].unique()[0]
    cat_1 = df[var_grupo].unique()[1]
    
    cols_numericas = df.select_dtypes(include=[np.number]).columns.tolist()
    for col in cols_numericas:
        if col != var_grupo:
            grupo_0 = df[df[var_grupo] == cat_0][col].dropna()
            grupo_1 = df[df[var_grupo] == cat_1][col].dropna()
            try:
                # Fisher: razón de varianzas F = var1/var2
                var_0 = np.var(grupo_0, ddof=1)
                var_1 = np.var(grupo_1, ddof=1)
                f_stat = var_0 / var_1
                df_0   = len(grupo_0) - 1
                df_1   = len(grupo_1) - 1
                p_value_varianza = 2 * min(
                    stats.f.cdf(f_stat, df_0, df_1),
                    1 - stats.f.cdf(f_stat, df_0, df_1)
                )
                if p_value_varianza >= alpha_varianza:
                    t_stat, p_value = stats.ttest_ind(grupo_0, grupo_1, equal_var=True)
                else:
                    t_stat, p_value = stats.ttest_ind(grupo_0, grupo_1, equal_var=False)
                rs_analisis.append([col, p_value_varianza, np.abs(t_stat), p_value])
            except Exception:
                continue
    t_stats = [(fila[0], fila[2]) for fila in rs_analisis]
    resultado_ordenado = sorted(rs_analisis, key=lambda x: x[2], reverse=True)
    headers = ['Variable', 'P-value Fisher', 'Estadística t', 'P-value']
    return tabulate(resultado_ordenado, headers=headers, floatfmt=".4f"), t_stats

**Fisher (igualdad de varianzas)**
$$H_0: S_1^2 = S_2^2$$
$$H_1: S_1^2 \neq S_2^2$$
$$F = \frac{S_1^2}{S_2^2} \sim F_{(n_1-1,\ n_2-1)}$$
* Si: $p < \alpha \Rightarrow \text{rechazar } H_0 \Rightarrow \text{varianzas distintas} \Rightarrow \text{Welch}$
* Si: $p \geq \alpha \Rightarrow \text{no rechazar } H_0 \Rightarrow \text{varianzas iguales} \Rightarrow \text{T-Student clásico}$

**T-Student (igualdad de medias)**
$$H_0: \mu_1 = \mu_2$$
$$H_1: \mu_1 \neq \mu_2$$
$$t = \frac{\bar{X}_1 - \bar{X}_2}{\sqrt{S_p^2\left(\frac{1}{n_1} + \frac{1}{n_2}\right)}} \sim t_{(n_1+n_2-2)}$$
* Si: $p < \alpha \Rightarrow \text{rechazar } H_0 \Rightarrow \mu_1 \neq \mu_2 \Rightarrow \text{variable discrimina}$
* Si: $p \geq \alpha \Rightarrow \text{no rechazar } H_0 \Rightarrow \mu_1 = \mu_2 \Rightarrow \text{variable no discrimina}$

In [115]:
sign_stats, _ = sign_estadistica(df, 'default', alpha_varianza=0.05)
print(sign_stats)

Variable       P-value Fisher    Estadística t    P-value
-----------  ----------------  ---------------  ---------
deu_ing                0.0096           7.3148     0.0000
a_emp                  0.0652           6.0728     0.0000
a_dir                  0.1195           2.7665     0.0059
ing_h                  0.2754           2.5573     0.0109
deu_o                  0.2283           2.3091     0.0214
deu_tc                 0.3553           2.2874     0.0226
edad                   0.1051           1.9856     0.0476
n_educ                 0.2630           1.8572     0.0639
flag_casado            0.9517           0.7500     0.4537
a_exp                  0.4264           0.3741     0.7085
id                     0.9092           0.3648     0.7154


* La variable que más discrmina los buenos y malos pagadores (`default = 1`) es la variable ratio `deuda_ingreso` con un `estadístico t = 7.3148`

**TAREA**

Cómo aplicarian el análisis de discriminación estadístico para hacer un filtro de variables mediante **coeficientes de correlación**, usando: 

1. Coef. correlación de Pearson (variables numéricas).
2. Coef. correlación de Cramer (variables categóricas).
3. Coef. correlación de Spearman (variables numéricas con alta concentración o asimetría).
4. Coef. correlación de Theil (variables categóricas con alta concentración o asimetría).
5. Coef. correlación de Tau de Kendall (variables numéricas con alta concentración o asimetría).

---
# 5. Modelamiento

Lograr entrenar un modelo estadístico o de machine learning con tal que logre extraer información de las variables dependientes sobre la variable independiente (target) para identificar patrones logrando estimar correctamente los buenos (y = 0) y malos pagadores (y = 1).

 $$ y = f(X) + \epsilon_i $$
 $$ y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_n X_n + \epsilon_i $$
 $$ default_i = \beta_0 + \beta_1 \cdot deuda\_ingreso + \beta_2 \cdot edad + ... + \beta_n \cdot X_n + \epsilon $$



## 5.1. Funciones de partidas

Se cargan:

* 1. Separación de variables para entrenamiento y validación
* 2. Función de Kolmogorov-Smirnov (KS).
* 3. Función iterativa para mrMR. 

1. Variables para **Entrenamiento** y **Validación**

In [116]:
y_ent = df_ent['default']
x_ent = df_ent.drop(columns=['default'])    

y_val = df_val['default']
x_val = df_val.drop(columns=['default'])

y_test = df_test['default']
x_test = df_test.drop(columns=['default'])

print("Dimensión Entrenamiento:", x_ent.shape)
print("Dimensión Validación:", x_val.shape)
print("Dimensión Test:", x_test.shape)
print("Variables de Entrenamiento:", x_ent.columns.tolist())
print("Variables de Validación:", x_val.columns.tolist())
print("Variables de Test:", x_test.columns.tolist())


Dimensión Entrenamiento: (233, 10)
Dimensión Validación: (100, 10)
Dimensión Test: (167, 10)
Variables de Entrenamiento: ['edad', 'n_educ', 'a_emp', 'a_dir', 'ing_h', 'deu_ing', 'deu_tc', 'deu_o', 'flag_casado_missing', 'a_exp_missing']
Variables de Validación: ['edad', 'n_educ', 'a_emp', 'a_dir', 'ing_h', 'deu_ing', 'deu_tc', 'deu_o', 'flag_casado_missing', 'a_exp_missing']
Variables de Test: ['edad', 'n_educ', 'a_emp', 'a_dir', 'ing_h', 'deu_ing', 'deu_tc', 'deu_o', 'flag_casado_missing', 'a_exp_missing']


2. Métrica a usar **Kolmogorov-Smirnov (KS)**.

In [117]:
# Cálculo de KS
def ks_statistic(y_true, y_pred_proba):
    fpr, tpr, _ = roc_curve(y_true, y_pred_proba)
    ks = np.max(tpr - fpr)
    return ks

3. **Método de MrMR** (Mínima Redundancia - Máxima Relevancia) - [Documentación de MrMR](https://feature-engine.trainindata.com/en/1.8.x/user_guide/selection/MRMR.html)

In [118]:
def ks_vars(estimator, x_ent, y_ent, x_val, y_val, max_vars=2, graph=False, subtitulo=''):
    ks_ent = []
    ks_val = []
    variable_names = []

    for num_vars in range(1, max_vars + 1):
        selected_features = mrmr_classif(x_ent, y_ent, K=num_vars)
        variable_names.append(list(selected_features))

        X_ent_selected = x_ent[selected_features]
        X_val_selected = x_val[selected_features]

        # Para modelos de statsmodels, agrega una constante
        if not is_classifier(estimator):
            X_ent_selected = sm.add_constant(X_ent_selected)
            X_val_selected = sm.add_constant(X_val_selected)
            model = estimator(y_ent, X_ent_selected)
            fitted_model = model.fit()
            y_pred_proba_ent = fitted_model.predict(X_ent_selected)
            y_pred_proba_val = fitted_model.predict(X_val_selected)

        else:
            # Para clasificadores de sklearn
            estimator.fit(X_ent_selected, y_ent)
            y_pred_proba_ent = estimator.predict_proba(X_ent_selected)[:, 1]
            y_pred_proba_val = estimator.predict_proba(X_val_selected)[:, 1]

        ks_ent_stat = ks_statistic(y_ent, y_pred_proba_ent)
        ks_val_stat = ks_statistic(y_val, y_pred_proba_val)

        ks_ent.append(ks_ent_stat)
        ks_val.append(ks_val_stat)

    if graph:
        import plotly.graph_objects as go
        fig = go.Figure()
        fig.add_trace(go.Scatter(x=np.arange(1, max_vars + 1), y=ks_ent, mode='lines+markers', name='Entrenamiento', line=dict(color='black', dash='solid')))
        fig.add_trace(go.Scatter(x=np.arange(1, max_vars + 1), y=ks_val, mode='lines+markers', name='Validación', line=dict(color='black', dash='dash')))
        fig.update_layout(title='Selección de variables usando KS',
                          xaxis_title='Número de variables',
                          yaxis_title='Estadístico KS',
                          legend_title='Conjunto',
                          width=1000, height=500,
                          template='simple_white',
                          title_text='Selección de Features usando Kolmogorov-Smirnov (KS)<br>'+subtitulo)
        fig.show()

    return ks_ent, ks_val, variable_names

## 5.2. Regresión Logística

La regresión logística es un modelo de clasificación binaria que estima la probabilidad de que un evento ocurra dado un conjunto de variables independientes. A diferencia de la regresión lineal, la variable dependiente es dicotómica (0/1), por lo que se utiliza la función sigmoide para acotar la salida entre 0 y 1.

**Modelo**
$$P(Y=1 \mid X) = \frac{1}{1 + e^{-(\beta_0 + \beta_1 X_1 + \beta_2 X_2 + \cdots + \beta_k X_k)}}$$

**Odds y Log-Odds** : ¿Cuántas veces más probable es que ocurra el evento en un grupo (P(Y=1)) versus otro (1 - P(Y=1))?
$$\text{Odds} = \frac{P(Y=1)}{1 - P(Y=1)}$$
$$\text{Logit} = \ln\left(\frac{P(Y=1)}{1-P(Y=1)}\right) = \beta_0 + \beta_1 X_1 + \cdots + \beta_k X_k$$

**Estimación de parámetros (Máxima Verosimilitud)** : Busca los valores de $\beta$ que maximizan la probabilidad de observar los datos que efectivamente ocurrieron.

$$\mathcal{L}(\beta) = \prod_{i=1}^{n} P(Y_i=1)^{y_i} \cdot (1 - P(Y_i=1))^{1-y_i}$$
$$\ell(\beta) = \sum_{i=1}^{n} \left[ y_i \ln P(Y_i=1) + (1-y_i) \ln(1-P(Y_i=1)) \right]$$

En esta sección revisaremos:

* 1. Ejecución del método de MrMR para selección de variables.
* 2. Entrenamiento del modelo de regresión logística.
* 3. Evaluación Metodologica, parte 1: métrica KS, F1, AUC, entre otras.
* 4. Evaluación Metodologica, parte 2: diagnostico de distribución de errores predictivos por variable.
* 5. Evaluación de Negocio.


**1. Ejecución del método de MrMR para selección de variables.**

In [123]:
max_vars = len(x_ent.columns.tolist())
logit_model = sm.Logit
ks_ent_vars_cl, ks_val_vars_cl, var_names = ks_vars(logit_model, x_ent, y_ent, x_val, y_val, max_vars=max_vars, graph=True
                                                                    ,subtitulo='Regresión Logística')

100%|██████████| 1/1 [00:00<00:00, 1861.65it/s]


Optimization terminated successfully.
         Current function value: 0.530876
         Iterations 6


100%|██████████| 2/2 [00:00<00:00,  9.34it/s]

Optimization terminated successfully.
         Current function value: 0.481501
         Iterations 6



100%|██████████| 3/3 [00:00<00:00, 17.92it/s]

Optimization terminated successfully.
         Current function value: 0.471190
         Iterations 6



100%|██████████| 4/4 [00:00<00:00,  6.93it/s]

Optimization terminated successfully.
         Current function value: 0.464002
         Iterations 7



100%|██████████| 5/5 [00:00<00:00, 10.14it/s]

Optimization terminated successfully.
         Current function value: 0.459661
         Iterations 7



100%|██████████| 6/6 [00:01<00:00,  4.37it/s]

Optimization terminated successfully.
         Current function value: 0.456679
         Iterations 7



100%|██████████| 7/7 [00:01<00:00,  5.83it/s]


Optimization terminated successfully.
         Current function value: 0.456623
         Iterations 7


100%|██████████| 8/8 [00:01<00:00,  4.73it/s]

Optimization terminated successfully.
         Current function value: 0.450765
         Iterations 7



100%|██████████| 9/9 [00:02<00:00,  4.25it/s]

         Current function value: 0.448764
         Iterations: 35



100%|██████████| 10/10 [00:02<00:00,  4.91it/s]

         Current function value: 0.447051
         Iterations: 35


**2. Entrenamiento del modelo de regresión logística.**

* Selección 1er Modelo: **8 variables seleccionadas por el método de MrMR**

In [124]:
# Asegúrate de que var_names está definido correctamente
selected_features = var_names[7]

# Añadiendo la constante al modelo
X_ent = sm.add_constant(x_ent[selected_features])
X_val = sm.add_constant(x_val[selected_features])
X_test = sm.add_constant(x_test[selected_features])

# Ajustando el modelo logístico
model = sm.Logit(y_ent, X_ent)
result = model.fit()

print("Variables seleccionadas:", selected_features)
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.450765
         Iterations 7
Variables seleccionadas: ['a_emp', 'deu_ing', 'a_dir', 'ing_h', 'edad', 'flag_casado_missing', 'n_educ', 'deu_o']
                           Logit Regression Results                           
Dep. Variable:                default   No. Observations:                  233
Model:                          Logit   Df Residuals:                      224
Method:                           MLE   Df Model:                            8
Date:                Sun, 03 May 2026   Pseudo R-squ.:                  0.2277
Time:                        22:13:01   Log-Likelihood:                -105.03
converged:                       True   LL-Null:                       -135.99
Covariance Type:            nonrobust   LLR p-value:                 1.954e-10
                          coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------

* Selección 2do Modelo: **3 variables seleccionadas por poder estadístico**

In [125]:
selected_features = var_names[3]

# Añadiendo la constante al modelo
X_ent = sm.add_constant(x_ent[selected_features])
X_val = sm.add_constant(x_val[selected_features])
X_test = sm.add_constant(x_test[selected_features])

# Ajustando el modelo logístico
model = sm.Logit(y_ent, X_ent)
result = model.fit()

print("Variables seleccionadas:", selected_features)
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.464002
         Iterations 7
Variables seleccionadas: ['a_emp', 'deu_ing', 'a_dir', 'ing_h']
                           Logit Regression Results                           
Dep. Variable:                default   No. Observations:                  233
Model:                          Logit   Df Residuals:                      228
Method:                           MLE   Df Model:                            4
Date:                Sun, 03 May 2026   Pseudo R-squ.:                  0.2050
Time:                        22:13:23   Log-Likelihood:                -108.11
converged:                       True   LL-Null:                       -135.99
Covariance Type:            nonrobust   LLR p-value:                 2.260e-11
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.4242      0.445     -3.197 

**3. Evaluación Metodologica, parte 1**

* El **AUC** (Area Under the Curve) de la curva ROC (Receiver Operating Characteristic)  
* La **Matriz de Confusión**.  
* El **KS** (Kolmogorov-Smirnov)  
* La **Precisión**, **Recall** y **F1-Score**  

In [126]:
# Predicciones de probabilidad
y_pred_proba_ent = result.predict(X_ent)
y_pred_proba_val = result.predict(X_val)
y_pred_proba_test = result.predict(X_test)

# Umbral óptimo (por defecto 0.5, pero puedes optimizarlo)
threshold = 0.5
y_pred_ent = (y_pred_proba_ent >= threshold).astype(int)
y_pred_val = (y_pred_proba_val >= threshold).astype(int)
y_pred_test = (y_pred_proba_test >= threshold).astype(int)

# F1-Score
f1_ent = f1_score(y_ent, y_pred_ent)
f1_val = f1_score(y_val, y_pred_val)
f1_test = f1_score(y_test, y_pred_test)

# KS
ks_ent = ks_statistic(y_ent, y_pred_proba_ent)
ks_val = ks_statistic(y_val, y_pred_proba_val)
ks_test = ks_statistic(y_test, y_pred_proba_test)

# AUC
auc_ent = roc_auc_score(y_ent, y_pred_proba_ent)
auc_val = roc_auc_score(y_val, y_pred_proba_val)
auc_test = roc_auc_score(y_test, y_pred_proba_test)

print("--F1-Score--")
print(f"F1-Score Entrenamiento: {f1_ent:.3f}")
print(f"F1-Score Validación: {f1_val:.3f}")
print(f"F1-Score Test: {f1_test:.3f}")
print("--KS--")
print(f"KS Entrenamiento: {ks_ent:.3f}")
print(f"KS Validación: {ks_val:.3f}")
print(f"KS Test: {ks_test:.3f}")
print("--AUC--")
print(f"AUC Entrenamiento: {auc_ent:.3f}")
print(f"AUC Validación: {auc_val:.3f}")
print(f"AUC Test: {auc_test:.3f}")

--F1-Score--
F1-Score Entrenamiento: 0.510
F1-Score Validación: 0.578
F1-Score Test: 0.339
--KS--
KS Entrenamiento: 0.516
KS Validación: 0.514
KS Test: 0.430
--AUC--
AUC Entrenamiento: 0.804
AUC Validación: 0.809
AUC Test: 0.731


* **Tarea**: Observar cómo cambian los resultados si se evaluan las métricas en vez de forma agregada separarlas por mes-año.

**4. Evaluación Metodologica, parte 2**: Diagnóstico de distribución de errores predictivos por variable 

Se realizó un análisis de densidad de variables por categoría de predicción para evaluar la capacidad de discriminación de las variables. El objetivo fue identificar en qué rangos de valor el modelo genera mayor certeza y en cuáles se producen los solapamientos que derivan en Falsos Positivos y Falsos Negativos.

In [127]:
df_ent['confusion_group'] = np.select(
    [
        (y_ent == 1) & (y_pred_ent == 1),  # TP
        (y_ent == 0) & (y_pred_ent == 0),  # TN
        (y_ent == 0) & (y_pred_ent == 1),  # FP
        (y_ent == 1) & (y_pred_ent == 0)   # FN
    ],
    ['TP', 'TN', 'FP', 'FN']
)

* Discriminación de `TP`, `TN`, `FP` y `FN` a través de gráficas de densidad por variables.

In [128]:
def plot_kde_by_confusion_group(
    df,
    features,
    group_col="confusion_group",
    order=("TN", "TP", "FP", "FN")
):
    var1, var2 = features[0], features[1]
    fig = make_subplots(rows=1, cols=2, subplot_titles=[var1, var2])
    colors = {'TN': 'blue', 'TP': 'green', 'FP': 'red', 'FN': 'orange'}
    for col_idx, var in enumerate([var1, var2], start=1):
        for g in order:
            data = df.loc[df[group_col] == g, var].dropna()
            kde_fig = ff.create_distplot(
                hist_data=[data],
                group_labels=[g],
                show_hist=False,
                show_rug=False,
                curve_type="kde"
            )
            fig.add_trace(
                go.Scatter(
                    x=kde_fig.data[0].x,
                    y=kde_fig.data[0].y,
                    name=g,
                    line=dict(color=colors[g]),
                    legendgroup=g,
                    showlegend=(col_idx == 1)  # leyenda solo en primer gráfico
                ),
                row=1, col=col_idx
            )
    fig.update_layout(
        title='Distribución de la variable por errores de clasificación',
        width=1200,
        height=600,
        paper_bgcolor='rgba(255,255,255,1)',
        plot_bgcolor='rgba(255,255,255,1)'
    )
    fig.update_yaxes(title_text='Densidad', col=1)
    fig.show()

In [129]:
plot_kde_by_confusion_group(df_ent, features=['deu_ing', 'edad'])

**Gráfico 1**

* Aquí se concentran los **TN** (azul) y los **FN** (rojo). Esto indica que cuando la relación deuda-ingreso es baja, el modelo tiende a predecir que el evento (default) no ocurrirá.

* Aquí se concentran los **TP** (naranja) y los **FP** (verde). Esto indica que cuando la deuda es alta respecto al ingreso, el modelo tiende a predecir que el evento (default) sí ocurrirá.

* **Nota.** Como hay un solapamiento significativo entre las 10 y 15 unidades de deu_ing, esa es la "zona de incertidumbre" donde el modelo comete la mayoría de sus errores de clasificación.

**5. Evaluación de Negocio.**

* **Paso 1.** De la muestra de Entrenamiento, se divide en 10 bucket la probabilidad estimada por el modelo.
* **Paso 2.** Se grafica cada uno de los 10 buckets vs la tasa de default observada en cada bucket promediando la variable `default`.
* **Paso 3.** Los mismo puntos de corte de cada bucket de la muestra de Entrenamiento, se aplican a la muestra de Validación y se grafica cada uno de los 10 buckets vs la tasa de default observada en cada bucket promediando la variable `default` pero en la muestra de Validación.

In [130]:
def distribucion_score(df_train, df_val, df_test, n_segments=10, score_col='SCORE', target_col='default', plot_graph=False):

    # Crear intervalos usando quantiles de df_train
    bins = pd.qcut(df_train[score_col], q=n_segments, retbins=True, duplicates='drop')[1]
    labels = [f"{round(bins[i], 2)} - {round(bins[i+1], 2)}" for i in range(len(bins)-1)]

    # Asignar intervalos a los tres DataFrames
    df_train = df_train.copy()
    df_val   = df_val.copy()
    df_test  = df_test.copy()
    df_train['Perfil'] = pd.cut(df_train[score_col], bins=bins, labels=labels, include_lowest=True)
    df_val['Perfil']   = pd.cut(df_val[score_col],   bins=bins, labels=labels, include_lowest=True)
    df_test['Perfil']  = pd.cut(df_test[score_col],  bins=bins, labels=labels, include_lowest=True)

    # Resúmenes
    resumen_train = df_train.groupby('Perfil').agg(
        Total=('Perfil', 'size'),
        Promedio_Default=(target_col, 'mean')
    ).reset_index()
    resumen_val = df_val.groupby('Perfil').agg(
        Total=('Perfil', 'size'),
        Promedio_Default=(target_col, 'mean')
    ).reset_index()
    resumen_test = df_test.groupby('Perfil').agg(
        Total=('Perfil', 'size'),
        Promedio_Default=(target_col, 'mean')
    ).reset_index()
    if plot_graph:
        # Gráfico entrenamiento
        fig_train = go.Figure()
        fig_train.add_trace(go.Bar(x=resumen_train['Perfil'], y=resumen_train['Total'], name='Total - Train', marker_color='indigo', yaxis='y1'))
        fig_train.add_trace(go.Scatter(x=resumen_train['Perfil'], y=resumen_train['Promedio_Default'], name='Probabilidad de Default - Train', marker=dict(color='orange', symbol='circle'), mode='lines+markers', yaxis='y2'))
        fig_train.update_layout(
            title='Entrenamiento',
            xaxis=dict(title='Perfil'),
            yaxis=dict(title='Total de Observaciones - Train', titlefont=dict(color='indigo')),
            yaxis2=dict(title='Probabilidad de Default - Train', titlefont=dict(color='orange'), overlaying='y', side='right'),
            legend=dict(orientation='h', yanchor='bottom', y=1.1, xanchor='center', x=0.5),
            width=1000, height=500, template='simple_white'
        )
        fig_train.show()
        # Gráfico validación
        fig_val = go.Figure()
        fig_val.add_trace(go.Bar(x=resumen_val['Perfil'], y=resumen_val['Total'], name='Total - Validation', marker_color='teal', yaxis='y1'))
        fig_val.add_trace(go.Scatter(x=resumen_val['Perfil'], y=resumen_val['Promedio_Default'], name='Probabilidad de Default - Validation', marker=dict(color='red', symbol='circle'), mode='lines+markers', yaxis='y2'))
        fig_val.update_layout(
            title='Validación',
            xaxis=dict(title='Perfil'),
            yaxis=dict(title='Total de Observaciones - Validation', titlefont=dict(color='teal')),
            yaxis2=dict(title='Probabilidad de Default - Validation', titlefont=dict(color='red'), overlaying='y', side='right'),
            legend=dict(orientation='h', yanchor='bottom', y=1.1, xanchor='center', x=0.5),
            width=1000, height=500, template='simple_white'
        )
        fig_val.show()
        # Gráfico test
        fig_test = go.Figure()
        fig_test.add_trace(go.Bar(x=resumen_test['Perfil'], y=resumen_test['Total'], name='Total - Test', marker_color='steelblue', yaxis='y1'))
        fig_test.add_trace(go.Scatter(x=resumen_test['Perfil'], y=resumen_test['Promedio_Default'], name='Probabilidad de Default - Test', marker=dict(color='crimson', symbol='circle'), mode='lines+markers', yaxis='y2'))
        fig_test.update_layout(
            title='Test',
            xaxis=dict(title='Perfil'),
            yaxis=dict(title='Total de Observaciones - Test', titlefont=dict(color='steelblue')),
            yaxis2=dict(title='Probabilidad de Default - Test', titlefont=dict(color='crimson'), overlaying='y', side='right'),
            legend=dict(orientation='h', yanchor='bottom', y=1.1, xanchor='center', x=0.5),
            width=1000, height=500, template='simple_white'
        )
        fig_test.show()
    return resumen_train, resumen_val, resumen_test

* Lo que buscamos en estas graficias que el modelo logre dismcriminar los mejores perfiles (menores bucket vs menores tasas de default observadas) a mayor probabilidad de incumplimiento observada, es decir, que la probabilidad de incumplimiento sea monotonicamente creciente a mayores niveles de buckets.

In [131]:
df_ent['SCORE_logistica'] = y_pred_proba_ent
df_val['SCORE_logistica'] = y_pred_proba_val
df_test['SCORE_logistica'] = y_pred_proba_test


dist_ent, dist_val, dist_test = distribucion_score(df_ent, df_val, df_test, n_segments=10, score_col='SCORE_logistica', plot_graph=True)

Podemos agrupar algunos perfiles (siempre observando gráfico de entrenamiento):

1. **Nuevo Perfil 1**: Perfil 1 con Perfil 2 (igual probabilidad).
2. **Nuevo Perfil 2**: Perfil 3 con Perfil 4 (para atenuar PD del perfil 3).
3. **Nuevo Perfil 3**: Perfil 5 con Perfil 6 
4. **Nuevo Perfil 4**: Perfil 7 con perfil 8.
5. **Nuevo Perfil 5**: Perfil 9.
6. **Nuevo Perfil 6**: Perfil 10.

observar cómo queda, y si no hay monotonicidad se pueden agrupar de otra forma. 

## 5.2. Random Forest Classifier

Random Forest es un algoritmo de aprendizaje de conjunto (ensemble) basado en la combinación de múltiples árboles de decisión. Cada árbol se entrena de forma independiente y el resultado final se obtiene por votación mayoritaria (clasificación) o promedio (regresión).

Se apoya en dos principios fundamentales:

1. **Bagging**: cada árbol se entrena con una muestra aleatoria con reemplazo del dataset original

2. **Aleatoriedad de variables**: en cada nodo, solo se evalúa un subconjunto aleatorio de variables para encontrar el mejor corte

En esta sección revisaremos:

* 1. Ejecución del método de MrMR para selección de variables.
* 2. Entrenamiento del modelo de Random Forest Classifier.
* 3. Evaluación Metodologica, parte 1: métrica KS, F1, AUC, entre otras.
* 4. Evaluación de Negocio.


**1. Ejecución del método de MrMR para selección de variables.**

In [152]:
max_vars = len(x_ent.columns.tolist())
rf_model = RandomForestClassifier(n_estimators=5, criterion = 'entropy', max_depth = 5, min_samples_split = 10, min_samples_leaf=10, random_state=42)
ks_ent_vars_rf, ks_val_vars_rf, var_names_rf = ks_vars(rf_model, x_ent, y_ent, x_val, y_val, max_vars=max_vars, graph=True, subtitulo='Random Forest')

100%|██████████| 10/10 [00:02<00:00,  4.77it/s]


* Los algoritmos de Random Forest utilizan la importancia para clasificiar las variables de mayor a menor importancia dentro de la descriminación de la variable default.

In [153]:
# Importancia de variables con Random Forest (barras verticales)
selected_features_rf = var_names_rf[4]  # Selecciona el mismo número de variables que en el ejemplo anterior

# Entrenamiento del modelo con las variables seleccionadas
rf_model_imp = RandomForestClassifier(n_estimators=10, criterion = 'entropy', max_depth = 5, min_samples_split = 10, random_state=42)
rf_model_imp.fit(x_ent[selected_features_rf], y_ent)

# Importancia de las variables
importancias = rf_model_imp.feature_importances_

df_importancias = pd.DataFrame({'Variable': selected_features_rf, 'Importancia': importancias})
df_importancias = df_importancias.sort_values(by='Importancia', ascending=False)
print(df_importancias)

# Gráfico de importancia (barras verticales)
fig = px.bar(df_importancias, x='Variable', y='Importancia', title='Importancia de Variables - Random Forest', color='Importancia', color_continuous_scale='Blues', orientation='v')
fig.update_layout(xaxis_title='Variable', yaxis_title='Importancia', bargap=0.2)
fig.show()

  Variable  Importancia
1  deu_ing     0.332477
0    a_emp     0.248710
3    ing_h     0.192860
2    a_dir     0.130727
4     edad     0.095226


**3. Evaluación Metodologica, parte 1: métrica KS, F1, AUC, entre otras.**

In [154]:
# Predicciones y métricas con Random Forest
y_pred_proba_ent_rf = rf_model_imp.predict_proba(x_ent[selected_features_rf])[:, 1]
y_pred_proba_val_rf = rf_model_imp.predict_proba(x_val[selected_features_rf])[:, 1]
y_pred_proba_test_rf = rf_model_imp.predict_proba(x_test[selected_features_rf])[:, 1]

# Umbral óptimo (por defecto 0.5, pero puedes optimizarlo)
threshold = 0.5
y_pred_ent_rf = (y_pred_proba_ent_rf >= threshold).astype(int)
y_pred_val_rf = (y_pred_proba_val_rf >= threshold).astype(int)
y_pred_test_rf = (y_pred_proba_test_rf >= threshold).astype(int)

# F1-Score
f1_ent_rf = f1_score(y_ent, y_pred_ent_rf)
f1_val_rf = f1_score(y_val, y_pred_val_rf)
f1_test_rf = f1_score(y_test, y_pred_test_rf)

# KS
ks_ent_rf = ks_statistic(y_ent, y_pred_proba_ent_rf)
ks_val_rf = ks_statistic(y_val, y_pred_proba_val_rf)
ks_test_rf = ks_statistic(y_test, y_pred_proba_test_rf)

# AUC
auc_ent_rf = roc_auc_score(y_ent, y_pred_proba_ent_rf)
auc_val_rf = roc_auc_score(y_val, y_pred_proba_val_rf)
auc_test_rf = roc_auc_score(y_test, y_pred_proba_test_rf)

print("--F1-Score--")
print(f"F1-Score Entrenamiento: {f1_ent_rf:.3f}")
print(f"F1-Score Validación: {f1_val_rf:.3f}")
print(f"F1-Score Test: {f1_test_rf:.3f}")

print("--KS--")
print(f"KS Entrenamiento: {ks_ent_rf:.3f}")
print(f"KS Validación: {ks_val_rf:.3f}")
print(f"KS Test: {ks_test_rf:.3f}")

print("--AUC--")
print(f"AUC Entrenamiento: {auc_ent_rf:.3f}")
print(f"AUC Validación: {auc_val_rf:.3f}")
print(f"AUC Test: {auc_test_rf:.3f}")

--F1-Score--
F1-Score Entrenamiento: 0.667
F1-Score Validación: 0.216
F1-Score Test: 0.259
--KS--
KS Entrenamiento: 0.819
KS Validación: 0.486
KS Test: 0.383
--AUC--
AUC Entrenamiento: 0.955
AUC Validación: 0.764
AUC Test: 0.709


* **Tarea**. Importante aplicar Grid Search para optimizar los hiperparámetros del modelo, como el número de árboles (`n_estimators`), la profundidad máxima de los árboles (`max_depth`), el número mínimo de muestras por hoja (`min_samples_leaf`), entre otros.

**4. Evaluación de Negocio.**

* Tiende a ordenar mejor los perfiles de riesgo vs probabilidad de incumplimiento observada, siendo que tienen las mismas variables que califican en el modelo final de regresión logística.

In [155]:
# Uso de la función para Random Forest
df_ent['SCORE_rf'] = y_pred_proba_ent_rf
df_val['SCORE_rf'] = y_pred_proba_val_rf
df_test['SCORE_rf'] = y_pred_proba_test_rf
dist_ent_rf, dist_val_rf, dist_test_rf = distribucion_score(df_ent, df_val, df_test, n_segments=10, score_col='SCORE_rf', plot_graph=True)

---
# 6. Tarea

* 1. Use otros algortimos: XGBoost, LightGBM, CatBoost, entre otros y compare los resultados con el modelo de regresión logística (**modelo benchmark**).

* 2. Seleccione un modelo, calcule: **(1).** Calibración y **(2).** Estabilidad de la probabilidad de incumplimiento estimada por el modelo usando PSI (Population Stability Index) entre la muestra de entrenamiento y validación.

* 3. Luego aplique caso de negocio en excel con las probabilidades entregadas por el modelo y los datos entregados del excel.

**Recuerde** que si selecciono un modelo tendrán que ahora ejecutar el mismo modelo con los mismo parametros pero con toda la muestra (sin separar entre entrenamiento y validación) para que el modelo aprenda de los datos que dejo fuera en la muestra de validación.

---